Initial import of "Device Code flow in .NET console app" (#24)
Azure-Samples/ms-identity-docs-code-dotnet: src/console-app-device-code/README.md

|---------------------------------:|:-------------------------------------------------------------------|:--------------------------------------------------------------------------------|
| **Name**                         | `dotnet-device-code-flow-app`                                      | Suggested value for this sample. <br/> You can change the app name at any time. |
| **Supported account types**      | **Accounts in this organizational directory only (Single tenant)** | Suggested value for this sample.                                                |
| **Platform type**                | _None_                                                             | No redirect URI required; don't select a platform.                              |
| **Allow public client flows**    | **Yes**                                                            | Required value for this sample.                                                 |

> :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

> ### 2. Update code sample in _Program.cs_ with app registration values

Reviewer: ```suggestion
### 2. Update the _Program.cs_ file with app registration values
```

ClientId = ""
```

## Run the application

```bash
dotnet run
```

Follow the device code flow instructions that are presented. If everything worked, you should receive a response similar to this:

```json
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#users/$entity",
  "businessPhones": ["+1 (999) 5551001"],
  "displayName": "Contoso Employee",
  "givenName": "Contoso",
  "jobTitle": "Worker",
  "mail": "cemployee@contoso.com",
  "mobilePhone": "1 999-555-1001",
  "officeLocation": "Contoso Plaza/F30",
  "preferredLanguage": null,
  "surname": "Employee",
  "userPrincipalName": "contoso_employee@contoso.com",
  "id": "e3a49d8b-d849-48eb-9947-37c1f9589812"
}
```

## About the code

This .NET 6 (C#) console application prompts the user to sign in via their device using a code provided by Microsoft Authentication Library (MSAL). The user completes this flow in their chosen web browser. Upon successful authentication, an HTTP GET request to the Microsoft Graph /me endpoint is issued with the user's access token in the HTTP header. The response from the GET request is then displayed in the console.

> This sample does not demonstrate access token caching. Access token caching should be used in situations where the console application will need to access the same protected API using the same access token multiple times across the life of the user's session in the application. This sample, as written, does not perform multiple calls and wouldn't benefit from a cache check. For additional information, see [Get a token from the token cache using MSAL.NET](https://docs.microsoft.com/azure/active-directory/develop/msal-net-acquire-token-silently).

Reviewer: If this were a CLI app, would I as the user be prompted to auth every time I issue a command that hits the API?

Author: I will say, you want disk token caching, there is no reason it should be limited to just this one sample, and we'll need to go back to all the others that were approved/merged and add it back in. (If you recall, we had the cache lookup in them and was asked to remove it because it wasn't ever being used -- they were all one-shot commands, no disk caching).

Reviewer: @ckittel Your first response was _beautiful_, Chad. I'm honestly not sure I've seen token caching explained that clearly before! We'll definitely leverage that in a how-to that could be a follow up this tutorial with that implements a persistent cache. :)
